executor/src/witgen/data_structures/finalizable_data.rs

```diff
@@ -0,0 +1,184 @@
+use std::{
+    collections::HashSet,
+    ops::{Index, IndexMut},
+};
+
+use ast::analyzed::PolyID;
+use bit_vec::BitVec;
+use number::FieldElement;
+
+use crate::witgen::rows::Row;
+
+/// A row entry in [FinalizableData].
+#[derive(Clone)]
+enum Entry<'a, T: FieldElement> {
+    /// The row is still in progress, and range constraints are still available.
+    InProgress(Row<'a, T>),
+    /// A finalized row, represented as a vector of values (corresponding to the columns
+    /// stored in [FinalizableData]) and a bit vector indicating which cells are known.
+    /// The value of unknown cells should be ignored.
+    Finalized(Vec<T>, BitVec),
+}
+
+/// A data structure that stores rows of a witness table, and behaves much like a `Vec<Row<T>>`.
+/// However, it also allows to finalize rows, which means that memory for things like range
+/// constraints is freed. The information which cells are known is preserved, though.
+/// Once a row has been finalized, any operation trying to access it again will fail at runtime.
+/// [FinalizableData::take_transposed] can be used to access the final cells.
+#[derive(Clone)]
+pub struct FinalizableData<'a, T: FieldElement> {
+    /// The list of rows (either in progress or finalized)
+    data: Vec<Entry<'a, T>>,
+    /// The list of column IDs (in sorted order), used to index finalized rows.
+    column_ids: Vec<PolyID>,
+}
+
+impl<'a, T: FieldElement> FinalizableData<'a, T> {
+    pub fn new(column_ids: &HashSet<PolyID>) -> Self {
+        Self::with_initial_rows_in_progress(column_ids, [].into_iter())
+    }
+
+    pub fn with_initial_rows_in_progress(
+        column_ids: &HashSet<PolyID>,
+        rows: impl Iterator<Item = Row<'a, T>>,
+    ) -> Self {
+        let mut column_ids = column_ids.iter().cloned().collect::<Vec<_>>();
+        column_ids.sort();
+        let data = rows.map(Entry::InProgress).collect::<Vec<_>>();
+        Self { data, column_ids }
+    }
+
+    pub fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    pub fn push(&mut self, row: Row<'a, T>) {
+        self.data.push(Entry::InProgress(row));
+    }
+
+    pub fn pop(&mut self) -> Option<Row<'a, T>> {
+        match self.data.pop() {
+            Some(Entry::InProgress(row)) => Some(row),
+            Some(Entry::Finalized(_, _)) => panic!("Row already finalized."),
+            None => None,
+        }
+    }
+
+    pub fn extend(&mut self, other: Self) {
+        self.data.extend(other.data);
+    }
+
+    pub fn remove(&mut self, i: usize) -> Row<'a, T> {
+        match self.data.remove(i) {
+            Entry::InProgress(row) => row,
+            Entry::Finalized(_, _) => panic!("Row {} already finalized.", i),
+        }
+    }
+
+    pub fn truncate(&mut self, len: usize) {
+        self.data.truncate(len);
```

Comment:
Shouldn't we use a finalized warning here as well? Is this used?

Reply:
Or is it OK since we do not return the removed data?

Reply:
Yes, exactly, the idea is that there is now way to get out a `Row<T>` after it has been finalized. But otherwise, you can still manipulate the array.